Add `grpc_ssl_certificate*` nginx config directives when using GRPCS …

S-Bohn opened the pull request:
Add `grpc_ssl_certificate*` nginx config directives when using GRPCS and `proxy-ssl-secret` annotation.
This allows using client-side mTLS authentication with backend servers.

<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:

It was not possible to use GRPCS backend protocol and enable client-side mTLS authentication (see https://github.com/kubernetes/ingress-nginx/issues/7082, https://github.com/kubernetes/ingress-nginx/issues/6893). This is caused by the fact that for gRPC nginx expects a specific configuration entry to specify the client certificate (see http://nginx.org/en/docs/http/ngx_http_grpc_module.html#grpc_ssl_certificate). 

The proposed workaround in the linked issues will effectively directly forward the connection on TCP level to the pod. This PR will generate the required settings in case a `proxy-ssl-secret` annotation was used and the protocol is gRPC. Which allows using mTLS with a backend server.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->
#6893
#7082 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Deployed an mTLS enabled gRPC server in a cluster. Used `grpcurl` with and without the change. Checked that without the change the server will terminate the connection. With the change, a client certificate is presented and the connection worked as expected.

I could not find an easy way to extend the e2e test suite, as it would require a grpcbin with enabled mTLS authentication.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

iamNoah1 reviewed (commented):
/triage accepted 
/priority important-soon
/kind bug

rikatz commented:
/ok-to-test
May we have some e2e scenario here? Thanks :)

iamNoah1 commented:
> /ok-to-test
> May we have some e2e scenario here? Thanks :)

+1 for adding new testcases @S-Bohn

S-Bohn commented:
> > /ok-to-test
> > May we have some e2e scenario here? Thanks :)
> 
> +1 for adding new testcases @S-Bohn

Let me try to extend grpcbin. 

I anyhow noticed that there are things missing: `grpc_ssl_verify` also exist. So please let me go through all `proxy-ssl-*` annotations and check whether they need a special gRPC case.

S-Bohn commented:
> News on this?

I updated it with an e2e test and added the missing `*_ssl_*` directives. I used helper functions to determine the backend protocol. Mainly required for the global per server case, where it could happen that both `grpc_*` and `proxy_*` directives are required. Not sure whether there is a better way.

I temporarily pushed a patched grpcbin version to get the tests running - that needs to be merged upstream first.

I would propose that I create a dedicated PR for the missing configuration entries that are different for gRPC: Probably a no-brainer for `grpc_buffer_size` and `grpc_*_timeout`. But I would need to double-check for `grpc_hide_header`, `grpc_intercept_errors`, and `grpc_next_upstream`.

James-Quigley commented:
Currently, if an ingress TLS Secret is used for `grpc_ssl_certificate*` and the Secret is updated (cert-manager rotation), the new secret value will not be passed to the backend server.

Not sure if this pull request can/should address this issue

visualphoenix commented:
+1 would love for this to get some attention

S-Bohn commented:
> +1 would love for this to get some attention

Yes, I need to follow up on this.

lsunsi commented:
@S-Bohn I just reached this PR coming from an issue exactly like #7082 . Have you found any workaround for this? Since both issues are closed and the PR is closed, I felt like I was crazy to not be able to get it working. But from the looks of it, it's sort of a "known issue" with mTLS with GRPC being literally unsupported?

lsunsi commented:
@S-Bohn Yeah, that's what I thought too. Do you remember what were the required directives? Would it be `grpc_ssl_certificate`, `grpc_ssl_certificate_key`, `grpc_ssl_server_name`, `grpc_ssl_name` and `grpc_ssl_verify`?

I mean, come to think of it, one alternative solution for nginx-ingress would be to just expose all of those via annotations right? It's worse than proposed solution because it's redundant, but it'd make it work instead of this just being unsupported.

joeycumines-scw commented:
Thank you for your efforts @S-Bohn! I was deeply unhappy with the fragile snippet kludge solution.

**I'm not sure if I'll actually be able to contribute (pending corporate CLA)** , but I've updated this PR, and refactored it a bit, adjusting the switching logic a little and unit testing it, and restructuring the templating a little.
These changes are available in branch `support-grpc-client-mtls-directives` of my public fork, if anyone is looking for an interim solution, largely based on this PR.

As an aside, my (manual) testing indicates that all seems to work as expected, though I wouldn't be surprised if there were some unhandled corner cases. I didn't bother setting up the E2E tests, so I've not tested those.
